tests/test_problem.py

```diff
@@ -0,0 +1,134 @@
+        BP.likelihood
+
+    with pytest.raises(ValueError, match=r"Unable to extract prior"):
+        BP.prior
+
+    with pytest.raises(ValueError, match=r"Unable to extract prior"):
+        BP.sample_prior(10)
+
+    with pytest.raises(ValueError, match=r"Unable to extract prior"): # Should rly be posterior
+        BP.UQ()
+
+    with pytest.raises(ValueError, match=r"Unable to extract prior"): # Should rly be posterior
+        BP.sample_posterior(10)
+
+    with pytest.raises(ValueError, match=r"Unable to extract prior"): # Should rly be posterior
+        BP.MAP()
+
+    with pytest.raises(ValueError, match=r"Unable to extract likelihood"):
+        BP.ML()
+
+    with pytest.raises(ValueError, match=r"Unable to extract likelihood"):
+        BP.data
+
+    with pytest.raises(ValueError, match=r"Unable to extract likelihood"):
+        BP.model
+
+def test_cannot_set_data():
+
+    x = cuqi.distribution.GaussianCov(0, 1)
+    y = cuqi.distribution.GaussianCov(lambda x: x, 1)
+
+    BP = cuqi.problem.BayesianProblem(y, x)
+
+    # Joint distribution does not care if we set the wrong data name (nothing happens)
+    # We add a check here to catch this in the future when its fixed.
+    BP.set_data(z=1) # z is not a valid name, should raise ValueError in future
+
+    # Set data
+    BP.set_data(y=1)
+
+    # Try to set data again
+    with pytest.raises(ValueError, match=r"data is already set"):
+        BP.set_data(y=1)
+
+def test_BayesianProblem_geometry_consistency():
+
+    domain_geometry = cuqi.geometry.StepExpansion(np.arange(5), 3)
+    range_geometry = cuqi.geometry.Continuous1D(10)
+
+    f = lambda x: np.hstack((x+5, x-5))
+    A = cuqi.model.Model(f, range_geometry, domain_geometry)
+    x = cuqi.distribution.GaussianCov(np.zeros(A.domain_dim), 1)
+    y = cuqi.distribution.GaussianCov(A, 1)
+    y_obs = y(x.sample()).sample() # Sample prior, then sample data distribution
+
+    # BayesianProblem
+    BP = cuqi.problem.BayesianProblem(x, y).set_data(y=y_obs)
+
+    # Check that we do not change the user provided geometries (We may want to change these in BP!)
+    # This is what the now removed _check_geometry_consistency did (but it did not do it for all cases)
+    # The tasks should be delegated to Posterior or JointDistribution
+    assert BP.posterior.prior.geometry == x.geometry # prior geometry is not changed
+    assert BP.posterior.likelihood.distribution.geometry == y.geometry # likelihood geometry is not changed
+    assert BP.data.geometry == y_obs.geometry # data geometry is not changed
+    assert BP.model == A # model geometry is not changed
+    
+    # Check that BayesianProblem correctly infers the geometry of the posterior (from model)
+    assert BP.posterior.geometry == domain_geometry
+    assert BP.sample_posterior(10).geometry == domain_geometry
+    #assert BP.sample_prior(10).geometry == domain_geometry # This fails because prior is sampled directly. TODO (should be domain_geometry)
```

Comment:
We mentioned in out discussion that we want to create an issue to capture geometries inconsistencies. Here is some documentation that can be converted to an issue.

For example:
- BP.sample_prior(10) generates samples with default geometry if prior has default geometry while the model has non-default geometry as domain geometry [sample prior should use domain geometry].
- Staked distribution: if two likelihood defined for the same prior, the samples generated from the stacked distribution will not be assigned the prior geometry and will be assigned default geometry instead [need to be assigned the prior geometry].

Reply:
#79